infra/src/Pos/Infra/Network/DnsDomains.hs

```diff
@@ -38,7 +62,111 @@ data NodeAddr a =
+                        ]
+                    (IPv6 (IP6 ip6), Just port) ->
+                        [ "addr" .= (A.String . show $ IPv6 (IP6 ip6))
+                        , "port" .= (toJSON port)
+                        ]
+                    (IPv6 (IP6 ip6), Nothing)   ->
+                        [ "addr" .= (A.String . show $ IPv6 (IP6 ip6))
+                        , "port" .= (toJSON (Prelude.read "3000" :: Word16))
+                        ]
+            (NodeAddrDNS nHost mWord) ->
+                case (nHost, mWord) of
+                    (Just host, Just port) ->
+                        [ "host" .= (toJSON $ decodeUtf8 @Text host)
+                        , "port" .= (toJSON port)
+                        ]
+                    (Just host, Nothing)   ->
+                        [ "host" .= (toJSON $ decodeUtf8 @Text host)
+                        , "port" .= (toJSON (Prelude.read "3000" :: Word16))
+                        ]
+                    (Nothing, _)      -> error "Please enter hostname"
+
+instance FromJSON (NodeAddr (Maybe DNS.Domain)) where
+    parseJSON (A.Object o)
+        | (HMS.member "addr" o) && (HMS.member "port" o) = do
+              addrV <- o .: "addr"
+              addr <- parseJSON addrV
+              portV <- o .: "port"
+              port <- parseJSON portV
+              return $ NodeAddrExact addr port
+        | (HMS.member "host" o) && (HMS.member "port" o) = do
+              portV <- o .: "port"
+              port <- parseJSON portV
+              hostV <- o .: "host"
+              case hostV of
+                  A.String text -> do
+                      let host = encodeUtf8 @Text @ByteString text
+                      return $ NodeAddrDNS (Just host) port
+                  _             -> fail "Incorrect JSON encoding for hostname"
+        | otherwise =
+            error "Incorrect JSON encoding for NodeAddr (Maybe DNS.Domain)"
+    parseJSON invalid = A.typeMismatch "NodeAddr (Maybe DNS.Domain)" invalid
+
+-- | Orphan instance; lifted from aeson-iproute
+instance ToJSON IP where
+    toJSON = A.String . T.pack . show
+
+-- | Orphan instance; lifted from aeson-iproute
+instance FromJSON IP where
+    parseJSON (A.String s)
+        | Just r <- readMaybe (T.unpack s) = pure r
+        | otherwise = fail "Unable to parse IP"
+    parseJSON v = A.typeMismatch "IP" v
+
+instance ToJSON (NodeAddr DNS.Domain) where
+    toJSON =
+        A.object . \case
+            (NodeAddrExact ip mWord) ->
+                 case (ip, mWord) of
+                     (IPv4 (IP4 ip4), Just port) ->
+                         [ "addr" .= (A.String . show $ IPv4 (IP4 ip4))
+                         , "port" .= (toJSON port)
+                         ]
+                     (IPv4 (IP4 ip4), Nothing)   ->
+                         [ "addr" .= (A.String . show $ IPv4 (IP4 ip4))
+                         , "port" .= (toJSON (Prelude.read "3000" :: Word16))
```

Comment:
@Jimbo4350 tells me this is a default value from [here](https://github.com/input-output-hk/cardano-sl/blob/8d25c2ad3ca2354af8f8c43a2972d1b9a31bf440/docs/network/topology-configuration.md#statically-configured-nodes-core-or-relay).

`(toJSON (Prelude.read "3000" :: Word16))` occurs 6 times in this file - we should DRY it up. I'd also suggest adding a comment with the link to the above topology-configuration document in a comment, so readers know where the value came from.

Reply:
Doing the above suggested refactoring will DRY it up to 3 occurrences, and those can be lifted out into a variable.

Reply:
Agreed, got lost in the fiddly JSON instances. I'll refactor :)